src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs

```diff
@@ -180,21 +189,27 @@ internal static bool ValidateSslServerCertificate(string targetServerName, X509C
                         if (certServerName[targetServerName.Length] != '.')
                         {
                             SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "targetServerName {0}, Target Server name does not match Subject in Certificate.", args0: targetServerName);
-                            return false;
+                            throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNameMismatch);
                         }
                     }
+#endif
                 }
-                else
+                else if ((policyErrors & SslPolicyErrors.RemoteCertificateChainErrors) != 0)
```

Comment:
Don't use else ifs here. They won't get hit correctly in the event of name plus other policy errors. Evaluate each policy error individually and then name mismatch last, which can have an else block for unknown policy errors.

You can also switch to policyErrors.HasFlag(x) in all these conditions to make the code easier to read.

We can also drill into the chain errors to provide more detail in the error. I'll send you some code suggestions offline.